b2sdk/encryption/setting.py

```diff
@@ -54,29 +97,78 @@ def __eq__(self, other):
         return self.mode == other.mode and self.algorithm == other.algorithm and self.key == other.key
 
     def as_value_dict(self):
+        """
+        Dump EncryptionSetting as dict for serializing a to json for requests.
+        """
+        if self.key and self.key.secret is None:
+            raise ValueError('cannot use an unknown key in transmission')
+        return self.repr_as_dict()
+
+    def repr_as_dict(self):
+        """
+        Dump EncryptionSetting as dict for representing.
+        """
         result = {'mode': self.mode.value}
         if self.algorithm is not None:
             result['algorithm'] = self.algorithm.value
+        if self.mode == EncryptionMode.SSE_C:
+            result.update(self.key.as_dict())
         return result
 
     def add_to_upload_headers(self, headers):
         if self.mode == EncryptionMode.NONE:
             # as of 2021-03-16, server always fails it
             headers['X-Bz-Server-Side-Encryption'] = self.mode.name
         elif self.mode == EncryptionMode.SSE_B2:
-            headers['X-Bz-Server-Side-Encryption'] = self.algorithm.name
+            self._add_sse_b2_headers(headers)
+        elif self.mode == EncryptionMode.SSE_C:
+            self._add_sse_c_headers(headers)
+            if self.key.key_id is not None:
+                header = 'X-Bz-Info-%s' % (SSE_C_KEY_ID,)
+                if headers.get(header) is not None and headers[header] != self.key.key_id:
+                    raise ValueError(
+                        'Ambiguous key id set: "%s" in headers and "%s" in %s' %
+                        (headers[header], self.key.key_id, self.__class__.__name__)
+                    )
+                headers[header] = self.key.key_id
```

Comment:
okay, so I think urllib.parse.quote and unquote will do the trick (instead of bytes) and testing with diacritic characters should be done in CLI integrations tests (which I have done just now)